### PR TITLE
Add option to project templates to use Program.Main instead of top-level statements

### LINKED ISSUE #40877: Add template option to generate explicit Program class with Main method instead of top-level statements
Related to dotnet/templating#3985

Add an option `--use-program-main` to the web and worker project templates that when specified generates a concrete `Program` class with a static `Main` method rather than using top-level statements.

- The option will default to disabled when used from the CLI
- When displayed in VS the option will be inverted and defaulted to enabled and will read:
    *Use top-level statements (uncheck to use an explicit Program class with a Main method)*

Potentially back-porting to 6.0.300 as well.

### PR DESCRIPTION
**Pull request by DamianEdwards**
Adds the `--use-program-main` option to the web project templates.

Spa templates will be updated separately.

Contributes to #40877

### CODE REVIEW

File: src/ProjectTemplates/test/BlazorServerTemplateTest.cs [hunk 1]
```diff
@@ -26,6 +26,9 @@ public BlazorServerTemplateTest(ProjectFactoryFixture projectFactory)
     [Fact]
     public Task BlazorServerTemplateWorks_NoAuth() => CreateBuildPublishAsync("blazorservernoauth");
 
+    [Fact]
+    public Task BlazorServerTemplateWorks_ProgamMainNoAuth() => CreateBuildPublishAsync("blazorservernoauth", args: new [] { "--use-program-main" });
```
Comment (HaoK) [inline src/ProjectTemplates/test/BlazorServerTemplateTest.cs:30]: I see mostly NoAuth variants, is auth not supported with --use-program-main?

Comment (DamianEdwards) [reply src/ProjectTemplates/test/BlazorServerTemplateTest.cs:30]: It's supported, I just didn't expand the test suite to cover every combination as it didn't appear to already. I can add more combinations if we think it's warranted.

**Comment by DamianEdwards** [comment]
Spa template updates: dotnet/spa-templates#42

**Review by HaoK — APPROVED** [review]
The changes to the project templates tests look good to me, but I didn't review all the templatestrings or json files.

**Comment by phenning** [comment]
@DamianEdwards Can we avoid inverting this boolean?  Instead default this to unchecked in Visual Studio? The templates should be deprecating the use of strings.json in favor of the built-in templatestrings.json. I know we have this currently for Open API, but we should consider that option grandfathered in.

If you keep it this way, it will impact the localization of this setting. NonENU langauges will display the ENU string. The localization for the strings.json does not live within the ASP.NET repo, and we are no longer round tripping these strings between dotnet/aspnetcore and WebTools. With that said, the aspnet repo could localize the strings.json in the repo and add it back into the templates from which it was previously removed.

You can also use "displayName" property in template.json as well.

Unfortunately there is not currently a way to override this in the template.json directlyu

It also seems rather wordy on screen - can any of this be moved to the description for the tooltip?

**Comment by DamianEdwards** [comment]
@phenning I've created #40917 to track un-inverting the property in VS.

**Pull request merged**